### PR TITLE
Make Plating honor 'self', and generalize how Klein deals with that.

### LINKED ISSUE #133: support for methods in klein.Plating
Right now, `Plating` breaks if it's used with an instance-level `Klein` object, since it expects that its first argument will be a `Request` object, and sometimes it's a bound instance.  This should be fixed.

### PR DESCRIPTION
**Pull request by glyph**
Fixes #133

### CODE REVIEW

File: src/klein/_decorators.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        to set attributes that will be visible at the top level.
+
+    @return: A new callable; this may have a different argument signature or
+        return value, and is only related to C{original} in the sense that it
+        likely calls it.
+    """
+    def decorator(wrapper):
+        result = (named(modification + ' for ' + original.__name__)
+                  (wraps(original)(wrapper)))
+        result.__original__ = original
```
Comment (moshez) [inline src/klein/_decorators.py]: should that be __klein_original__? or is __original__ a standard?

File: src/klein/_decorators.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        if modifier is not None:
+            before = set(wrapper.__dict__.keys())
```
Comment (moshez) [inline src/klein/_decorators.py]: there are a lot of assumptions here about wrapper (e.g., that stuff is in `__dict__`). dir() has garbage in it, but the garbage should be the same before and after, so perhaps using dir(), setattr and getattr make more sense?

File: src/klein/_decorators.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        return result
+    return decorator
+
+
+def named(name):
+    """
+    Change the name of a function to the given name.
+    """
+    def decorator(original):
+        original.__name__ = str(name)
```
Comment (moshez) [inline src/klein/_decorators.py]: should that be `__klein_name__`? for regular functions, the real name is in func_name, right?

File: src/klein/_plating.py [hunk 4]
```diff
@@ -96,54 +114,58 @@ def __init__(self, defaults=None, tags=None,
...
-            def mymethod(request, *args, **kw):
-                data = method(request, *args, **kw)
+            loader = TagLoader(tags)
+            @modified("plating route renderer", method, routing)
+            @bindable
+            @inlineCallbacks
+            def mymethod(instance, request, *args, **kw):
+                data = yield _call(instance, method, request, *args, **kw)
                 if _should_return_json(request):
                     json_data = self._defaults.copy()
```
Comment (moshez) [inline src/klein/_plating.py]: I'm seeing these 4 lines:

```
json_data = some_dict.copy()
for ignored in self._presentationSlots:
     json_data.pop(ignored_None)
return json_data
```

Pop up like 3 times. Sounds like this should be a method (`with_presentation_removed` maybe?)

File: src/klein/_decorators.py (hunk 3 above)
Comment (glyph) [inline src/klein/_decorators.py]: No, that was an outdated 2.x-ism.

```python
Python 2.7.13 (default, Dec 22 2016, 19:17:50) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)] on darwin
>>> def foo(): pass
... 
>>> foo.__name__
'foo'
>>> 
Python 3.6.0 (default, Mar 11 2017, 01:25:45) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)] on darwin
>>> def foo(): pass
... 
>>> foo.__name__
'foo'
>>> 
```

File: src/klein/_decorators.py (hunk 1 above)
Comment (glyph) [inline src/klein/_decorators.py]: `__klein_original__` you mean? I have pretensions for this code to eventually be more general ;)

File: src/klein/_plating.py (hunk 4 above)
Comment (glyph) [inline src/klein/_plating.py]: Thanks for spotting that, 

NB: Klein has been a bit inconsistent in following its own documentation, but [it should be using Twisted coding standard names](https://github.com/twisted/klein/blob/d60e0bf567f858201c41399c2c7750493911cf74/CONTRIBUTING.rst#code) so in the future, `withPresentationRemoved` :)

**Comment by glyph** [comment]
Will merge after fixing the CI failure oops :)

File: src/klein/_decorators.py (hunk 3 above)
Comment (glyph) [inline src/klein/_decorators.py]: And: no, this definitely shouldn't be `__klein_name__` - the whole point is to get it to show up in the `repr` :).  (For which there is a test.)`

**Comment by glyph** [comment]
Looks like the reported coverage issues are real, I'll hopefully address those today.

**Pull request merged**